File: Tutorial4-6SubComponentBuildersMultipleModulesAndActvities/src/main/java/com/example/tutorial4_6subcomponentbuildersmultiplemodulesandactvities/di/ApplicationComponent.java
package com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.di;


import android.app.Application;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;

// This is parent component
@Component(modules = ApplicationModule.class)
@Singleton
public interface ApplicationComponent {

    /**
     * Calls SubComponent Builder in SecondActivity and MyFragment using ApplicationComponent
     */
    ToastAndPreferencesSubComponent.Builder toastMakerBuilder();

    DummyDependencyComponent.Builder dummyDependencyBuilder();

    @Component.Builder
    interface Builder {

        @BindsInstance
        Builder application(Application application);

        ApplicationComponent build();

    }
}

File: Tutorial4-6SubComponentBuildersMultipleModulesAndActvities/src/main/java/com/example/tutorial4_6subcomponentbuildersmultiplemodulesandactvities/di/ApplicationComponent.kt
package com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.di

import android.app.Application
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

// This is parent component
@Component(modules = [ApplicationModule::class])
@Singleton
interface ApplicationComponent {

    /**
     * Calls SubComponent Builder in SecondActivity and MyFragment using ApplicationComponent
     */
    fun toastMakerBuilder(): ToastAndPreferencesSubComponent.Builder

    fun dummyDependencyBuilder(): DummyDependencyComponent.Builder

    @Component.Builder
    interface Builder {

        @BindsInstance
        fun application(application: Application): Builder

        fun build(): ApplicationComponent
    }
}